### ISSUE TITLE
Github Branching Plan

### ISSUE THREAD

**Issue by CharlieHewitt** [CharlieHewitt/IndustrialTeamProject#3]
## GitHub Structure

The default branch is _dev_. We will be working on features in groups or as individuals on _feature branches_ . You should name these feature branches following this naming pattern: _teamName-featureName_ (eg backend-db-connection)

![image](https://user-images.githubusercontent.com/33226685/95080907-b1df3a80-0710-11eb-9312-f6a12ebab322.png)

Any work on that feature should take place on that branch. You can push the feature branch to GitHub if you need to share it with other people and/or the feature isn't finished yet. Please commit and push/pull often when working, as this will minimise the number of problems we will run into.

```
// creating a feature branch:
git branch (check which branch you are on)
git checkout dev (optional, swap to dev branch if not on it already)
git pull (get latest changes on dev branch)
git checkout -b branch-name (create and swap to new branch)

// when working
git add (files to be commited or . for all changes)
git commit -m "useful commit message"
git push (push feature branch to the repository) 
```

## Pull Requests and Code Review

Once you have finished work on the feature, open a pull request on GitHub to merge the branch into _dev_. This pull request will have to be reviewed by 2 people for it to be successfully merged. Please do take the time to review things properly, code reviews are one of the best ways to look at how other people code and improve. If you don't understand what somebody has done? or how it works? don't blindly approve it, You can ask them how it works and you will hopefully have learnt something new!

When creating pull requests, use a descriptive name, and mention the issue it refers to, this will help us manage the project as whole. You can mention issues by typing # -> a list of issues pops up.

## Travis and Testing (CI)

Once Travis is up and running, each pull request will automatically run all the unit tests we have. If one of these tests fails, this needs to be looked at. (Either there is a bug in the code, or the test needs to be updated)

Try not to rely on Travis to run tests though! You can run them locally as well, I'll set up the option to have a terminal window open running our node tests in the background tomorrow (Tuesday 6/10).

I will also put a post up as a guide to how to write/add unit tests for code you are writing.

## TLDR

- main branch is _dev_
- use feature branches for (preferably small) features!
- pull, commit and push often!
- when a feature is finished
    - create a pull request (descriptive name + mention issue (#) )
    - pull request reviewed by 2 people
    - update project board/issues!!
- use project boards and create issues for features/bugs/things that have to be done

## Questions/ Problems

If you have any questions or if something is unclear in the post above let me know! I'll update it.
If you run into any issues with anything GitHub related let me know and I will try my best to help.
